icu4c/source/i18n/messageformat2_function_registry.cpp

```diff
@@ -414,25 +416,185 @@ static FormattedPlaceholder notANumber(const FormattedPlaceholder& input) {
     return FormattedPlaceholder(input, FormattedValue(UnicodeString("NaN")));
 }
 
-static FormattedPlaceholder stringAsNumber(const number::LocalizedNumberFormatter& nf, const FormattedPlaceholder& input, UErrorCode& errorCode) {
+// Returns true if `c` is in the interval [`first`, `last`]
+static bool inRange(UChar32 c, UChar32 first, UChar32 last) {
+    U_ASSERT(first < last);
+    return c >= first && c <= last;
+}
+
+static bool isDigit(UChar32 c) { return inRange(c, 0x0030, 0x0039); }
+
+static int32_t parseDigit(UChar32 c) {
+    switch(c) {
+    case DIGIT_ZERO: {
+        return 0;
+    }
+    case DIGIT_ONE: {
+        return 1;
+    }
+    case DIGIT_TWO: {
+        return 2;
+    }
+    case THREE: {
+        return 3;
+    }
+    case FOUR: {
+        return 4;
+    }
+    case FIVE: {
+        return 5;
+    }
+    case SIX: {
+        return 6;
+    }
+    case SEVEN: {
+        return 7;
+    }
+    case EIGHT: {
+        return 8;
+    }
+    case NINE: {
+        return 9;
+    }
+    default: {
+        // Should be unreachable
+        return 0;
```

Comment:
```suggestion
    default: {
        // Should be unreachable
        U_ASSERT(isDigit(c));
        return 0;
```

should this handle other numbering systems though?

Reply:
and not to bikeshed on this part but there is `ufmt_digitvalue()` and you could ignore values <0 or ≥10 (it parses hex)

Reply:
> should this handle other numbering systems though?

No, because it's operating on a Number Operand as defined in [the function registry spec](https://github.com/unicode-org/message-format-wg/blob/main/spec/registry.md#number-operands); it's not a localized number.

> and not to bikeshed on this part but there is ufmt_digitvalue() and you could ignore values <0 or ≥10 (it parses hex)

I tried just now to use that, but I realized that `ufmt_digitvalue()` is part of the `io` library and `io` depends on `i18n`, so that would be a circular dependency.  I guess it could be moved into `i18n`, but I'm hesitant to do a refactor like that just to avoid duplicating a relatively small function. What do you think?

Reply:
I didn't realize it was in io, good call. fine as is then.